### ISSUE TITLE
[Idea] allow sending arguments together with method name for reindexing

### ISSUE THREAD

**Issue by strzibny** [ankane/searchkick#1148]
Hi,

currently we can send a method name for re-indexing:
```
Record.reindex(:partial_search_data)
```
I would like to be able to make the method dynamic based on input so that I could pass an argument (like an associated record ID):
```
Record.reindex(:partial_search_data, 20)
```
The only alternative I can think of is to be sending a method like `partial_search_data_20` and use meta programming (`method_missing`) to catch and build it but I don't think it's elegant.

I also looked into code and seems like accepting params and just passing it might not be too much work (meaning I can attempt to try to implement it)? Let me know if you would accept a PR like this.

**Comment by strzibny** [comment]
@ankane Thank you for a quick reply. I appreciate it.

I will try to present you my use-case in more detail and maybe it becomes a bit more clear.

Imagine you have a Document and File model where one document can have `0..n` files. Now all those files might be added for document full-text search by getting the text data from them (think OCR, or just reading a text file) and sending to Elasticsearch.

With current Searchkick I am sure able to iterate over all associated files (in `partial_search_data` method above), getting their text layer for full-text search (if already available) and sending it. This can be very expensive performance wise. Even though Elasticsearch will have to do the work anyway (by partial reindexing replacing the whole document) I still believe I can save a lot of processing power to partially reindex only a new added file in the document (or a file which content we recognized with OCR).

Basically I want to send only new data to ES and not iterate over 50 files and getting the data from all of them. I am new to Elasticsearch and perhaps I am looking at the problem in the wrong way and there is different solution. If I can tell Searchkick send only data from file number 25 that be much nicer.

What do you think?

**Comment by mikelkew** [comment]
@strzibny If @ankane doesn't want to proceed with this, perhaps another approach for this is to scope the associated files being referenced in your `partial_search_data` method to those that have been updated more recently than the parent Record. Unless you are touching Record on the association this should be fairly easy.

If you _are_ touching the parent, the other somewhat fiddly approach would be to ensure the updated_at field is being indexed, and then scope the associated files in your method by comparing their timestamps with the updated_at field in the currently indexed document, e.g.:
```ruby
doc = Record.search_index.retrieve(self).presence
files_to_index = doc ? associated_files.where('updated_at > ?', Time.parse(doc.updated_at) : associated_files
```

**Comment by ankane** [comment]
@strzibny I understand the use case, but it still seems uncommon. The `method_missing` approach you mentioned above should work fine.

Also, as a general rule, I'd avoid doing any heavy lifting in `search_data` methods. If you need to do an emergency reindex, this will make it painful. Instead, I'd cache the results of the OCR, and only read those results when reindexing.

**Comment by strzibny** [comment]
@mikelkew Thanks, interesting approach. I might try it.

@ankane 

> Instead, I'd cache the results of the OCR, and only read those results when reindexing.

I actually didn't say I don't want to do the heavy lifting beforehand. However it is not the solution to avoid iterating over many records (think 50, 100), reading all of this data into memory and sending to ES. Doing this on every added file means it might be done those 50 times (so 50 times iterating over 1..50 records, getting all their big text layer and sending). I just don't like how wasteful it is:).

I really think it's a very small and non-breaking change to Searchkick enabling a lot of useful and more performant use-cases so I will probably attempt to do a patch myself that I will have to maintain on top of Searchkick. If that fails I might need to do this directly outside Searchkick.


 I respect that you don't want to accept a change like this and thanks for the giving it a though.
